### PR TITLE
Openai interface

### PR DESCRIPTION
**Pull request by Rowing0914**
Hi Mads,

As we talked on LinkedIn, I would like to push the openai interface to this repo!
To test it, you can try it from `Source/main_openai.py`(https://github.com/Rowing0914/PyBoy/blob/openai_interface/Source/main_openai.py)!!

Thanks

### CODE REVIEW

**Comment by krs013** [comment]
Heads up @Rowing0914, your commits have the Cythonized `.c` files in it, which need to be removed before anything could be merged. Also, it looks like at least some of the files you have are based on older commits, notably `pyboy.py`, so you might need to revisit what you did to merge upstream changes while working. Both of these things will probably have to be fixed before Mads can do much to test and merge it, otherwise he has to do all the work on his end without knowing what of your patch is part of the patch or just an accident. Feel free to comment here or visit the Discord channel if you want to follow up on any of this!

File: Source/main.py [hunk 1]
```diff
@@ -7,8 +7,8 @@
 import os
 import traceback
 import sys
-from pyboy import PyBoy
-from pyboy.logger import addconsolehandler
+from Source.pyboy import PyBoy
```
Comment (Baekalfen) [inline Source/main.py]: I looks like you are importing the module from the wrong directory. If you need to use PyBoy from another code base, then install the code with `make install`.

File: README.md [hunk 2]
```diff
@@ -26,6 +26,34 @@ The code base is still pure Python, which means it will still run in CPython and
 At the Wiki page, you will also find out how to interface with PyBoy from your own project: [Wiki](https://github.com/Baekalfen/PyBoy/wiki).
 
 
+
+# OpenAI Interface
```
Comment (Baekalfen) [inline README.md]: This should be moved to the Wifi when we merge, but for now, we can leave it here.

File: Source/main.py [hunk 3]
```diff
@@ -54,7 +54,7 @@ def main():
             filename = getROM(romdir)
 
         # Start PyBoy and run loop
-        pyboy = PyBoy(sys.argv[1] if len(sys.argv) > 1 else None, scale, filename, bootrom)
+        pyboy = PyBoy(sys.argv[1] if len(sys.argv) > 1 else None, filename, scale, bootrom)
```
Comment (Baekalfen) [inline Source/main.py]: We cannot change the interface without warning, as everyone else using PyBoy will then have to change their code.

File: Source/pyboy/pyboy.py [hunk 4]
```diff
@@ -119,11 +126,25 @@ def stop(self, save=True):
...
             print("Profiling report:")
             from operator import itemgetter
-            names = [CPU_COMMANDS[n] if n < 0x100 else CPU_COMMANDS_EXT[n-0x100] for n in range(0x200)]
+            names = [CPU_COMMANDS[n] if n < 0x100 else CPU_COMMANDS_EXT[n - 0x100] for n in range(0x200)]
             for hits, n, name in sorted(
                     filter(itemgetter(0), zip(self.mb.cpu.hitRate, range(0x200), names)), reverse=True):
                 print("%3x %16s %s" % (n, name, hits))
 
+    def step(self, action):
+        # === Internal Process to move one step forward ===
```
Comment (Baekalfen) [inline Source/pyboy/pyboy.py]: Won't it be better to call the `.tick()` method?

File: Source/pyboy/pyboy.py [hunk 5]
```diff
@@ -171,3 +192,13 @@ def set_emulation_speed(self, v, max_speed=0):
         if max_speed > 5:
             logger.warning("The emulation speed might not be accurate when higher than 5")
         self.max_emulationspeed = max_speed
+
+    def _readout_frame(self, frame):
+        """ Readout a raw frame of the GameBoy Emulator """
+        return Image.frombytes(self.mb.window.color_format, (SC_HEIGHT, SC_WIDTH), frame)
```
Comment (Baekalfen) [inline Source/pyboy/pyboy.py]: This could just be inlined in `get_obs`.

File: Source/pyboy/pyboy.py [hunk 6]
```diff
@@ -171,3 +192,13 @@ def set_emulation_speed(self, v, max_speed=0):
...
+
+    def get_obs(self):
+        """ Convert the Raw frame into Numpy array and return it """
+        raw_frame = self._readout_frame(self.getScreenBuffer())
+        data = np.asarray(raw_frame, dtype="int32")
```
Comment (Baekalfen) [inline Source/pyboy/pyboy.py]: If OpenAI expects a specific format of the frame buffer, I would prefer to have it in a separate wrapper class/file.

File: Source/pyboy/pyboy.py [hunk 7]
```diff
@@ -119,11 +126,25 @@ def stop(self, save=True):
...
+        if not self.paused:
+            self.mb.tickframe() # TODO: I don't know what this does...
+        self.window.update_display() # update the screen
+        # === Internal Process done ===
+
+        observation = self.get_obs()
+        reward = -1.0  # TODO: think the reward design
+        done = False # TODO: when to done??
+        info = {}
+        return observation, reward, done, info
```
Comment (Baekalfen) [inline Source/pyboy/pyboy.py]: If OpenAI expects a specific format for the step function, I would prefer to have it in a separate wrapper class/file.

**Comment by Rowing0914** [comment]
Thanks!
since I have been busy with my thesis for the moment, I will deal with this after the submission!
Best,

**Comment by Baekalfen** [comment]
@Rowing0914 Sounds good! I completely understand :)

**Comment by Baekalfen** [comment]
Closed for inactivity

**Comment by Rowing0914** [comment]
Sorry for leaving inactive.,..
I will merge the latest master branch again and come back here to finish this PR soon..
It seems I'll be able to land on a full-time job soon!

**Comment by Baekalfen** [comment]
Don't worry about it. Someone else started working on it in #126

**Comment by Rowing0914** [comment]
> Don't worry about it. Someone else started working on it in #126

Got it! Thank you for the reference!
I will check it!

**Pull request closed without merging**